jvb/src/main/kotlin/org/jitsi/videobridge/shutdown/ShutdownManager.kt

```diff
@@ -0,0 +1,106 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.videobridge.shutdown
+
+import org.jitsi.meet.ShutdownService
+import org.jitsi.utils.logging2.Logger
+import org.jitsi.utils.logging2.createChildLogger
+import org.jitsi.videobridge.shutdown.ShutdownConfig.Companion.config
+import org.jitsi.videobridge.shutdown.ShutdownState.GRACEFUL_SHUTDOWN
+import org.jitsi.videobridge.shutdown.ShutdownState.RUNNING
+import org.jitsi.videobridge.shutdown.ShutdownState.SHUTTING_DOWN
+import org.jitsi.videobridge.util.TaskPools
+import java.util.concurrent.TimeUnit
+
+/**
+ * Implements the graceful shutdown logic.
+ *
+ * When graceful shutdown is initiated the state changes to [GRACEFUL_SHUTDOWN], and this is advertised in presence.
+ * In this mode new conferences are rejected, but existing conferences continue to run and accept new participants. We
+ * stay in this state until either [config.gracefulShutdownMaxDuration] has passed, or the number of participants
+ * drops below [config.gracefulShutdownMinParticipants].
+ *
+ * Then the state transitions to [SHUTTING_DOWN], and this state is advertised in presence. We expect jicofo to
+ * actively move participants off this bridge. We stay in [SHUTTING_DOWN] for [config.shuttingDownDelay] before we
+ * actually perform the shutdown.
+ */
+class ShutdownManager(
+    private val shutdownService: ShutdownService,
+    parentLogger: Logger
+) {
+    val logger = createChildLogger(parentLogger)
+
+    var state: ShutdownState = RUNNING
+        private set
+
+    fun initiateShutdown(graceful: Boolean) {
+        logger.info("Received shutdown request, graceful=$graceful")
+
+        if (graceful) {
+            if (state == RUNNING) {
+                state = GRACEFUL_SHUTDOWN
+                logger.info(
+                    "Entered graceful shutdown mode, will stay in this mode for up to " +
+                        config.gracefulShutdownMaxDuration
+                )
+                TaskPools.SCHEDULED_POOL.schedule(
+                    {
+                        if (state == GRACEFUL_SHUTDOWN) {
+                            logger.info("Graceful shutdown period expired, changing to SHUTTING_DOWN.")
+                            changeToShuttingDown()
+                        }
+                    },
+                    config.gracefulShutdownMaxDuration.toMillis(),
+                    TimeUnit.MILLISECONDS
+                )
+            } else {
+                logger.info("Graceful shutdown already initiated mode.")
+            }
+        } else {
+            changeToShuttingDown()
```

Comment:
If I'm reading this correctly, the non-graceful shutdown time has now gone from 1 second to 1 minute?  That's a pretty big operational difference.

Reply:
That's correct.

Graceful shutdown was already delayed by 1 minute (even when empty) to assure presence makes it to jicofo. I thought the same shoud apply to non-graceful shutdown and the code was simpler that way. I don't actually know why non-graceful shutfdown exists, I haven't seet it used anywhere.